src/historywork/DownloadBucketsWork.cpp

```diff
@@ -79,4 +84,41 @@ DownloadBucketsWork::yieldMoreWork()
     ++mNextBucketIter;
     return w3;
 }
+
+void
+DownloadBucketsWork::logArchivesForBuckets() const
+{
+    if (mArchive)
+    {
+        CLOG(INFO, "History")
+            << fmt::format("Buckets from archive {}", mArchive->getName());
+    }
+    else
+    {
+        // It's okay to use INFO here, as we only get up to 31 buckets
+        for (auto const& pair : mDownloads)
+        {
+            auto ar = pair.second->getArchive();
+            if (ar)
+            {
+                CLOG(INFO, "History") << fmt::format(
+                    "Bucket {} from archive {}", pair.first, ar->getName());
+            }
+        }
+    }
+}
+
+void
+DownloadBucketsWork::onSuccess()
+{
+    // Log archive regardless of success or failure in case downstream works
+    // (such as apply buckets) fail with corrupt buckets
+    logArchivesForBuckets();
+}
+
+void
+DownloadBucketsWork::onFailureRaise()
+{
+    logArchivesForBuckets();
```

Comment:
On failure, it would make a lot more sense to follow the pattern that you're using for `DownloadApplyCheckpoint` : basically pass a special "context" callback to `VerifyBucketWork` that way you can log the archive that was used in the context of the failure (verification failure).

Otherwise the logs will contain a dump of (potentially 31) archives where one them is faulty.